# # Capstone Final project

### Analysis of restaurants in Colombo, Sri Lanka neighbourhood to advise restaurant owners on the neighbourhood to open new restaurants

* Build a dataframe of neighborhoods in Colombo SriLanka by web scraping the data from Wikipedia page
* Get the geographical coordinates of the neighborhoods
* Obtain the venue data of the neighborhoods from Foursquare API
* Explore and cluster the neighborhoods

Import libraries

In [24]:

import numpy as np # library to handle arrays

import pandas as pd # import pandas library as pd
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

# library to handle JSON files
import json 

# convert address into latitude and longitude values
from geopy.geocoders import Nominatim 

!conda install -c conda-forge geocoder
import geocoder 

# library to handle requests
import requests 

# library for web scraping
from bs4 import BeautifulSoup 

# tranform JSON file into a pandas dataframe
from pandas.io.json import json_normalize 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes

# plotting library
import folium

print('Folium installed')
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          59 KB

The following NEW packages will be INSTALLED:

    geocoder: 1.38.1-py_1 conda-forge
    ratelim:  0.1.6-py_2  conda-forge


ratelim-0.1.6        | 6 KB      | ##################################### | 100% 
geocoder-1.38.1      | 53 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done

# All requested packages alread

In [28]:
data = requests.get('https://en.wikipedia.org/wiki/Category:Suburbs_of_Colombo').text
# webscrape wikipedia page using beautifulsoup object
soup = BeautifulSoup(data, 'html.parser')
neighborhoodList = []
# append the data into the list
for row in soup.find_all("div", class_="mw-category")[0].findAll("li"):
    neighborhoodList.append(row.text)

In [29]:
# create a new DataFrame from the list
col_df = pd.DataFrame({"Neighborhood": neighborhoodList})
col_df.head()

Neighborhood
0   Athurugiriya
1  Bambalapitiya
2   Battaramulla
3      Batuwatta
4    Bloemendhal

In [30]:
# print the number of rows of the dataframe
col_df.shape

(65, 1)

Latitude and Longitude of Colombo

In [38]:
# function to get latitude and longitude of Colombo
def get_latlng(neighborhood):
    coords = None
    while(coords is None):
        geo = geocoder.arcgis('{}, Colombo, SriLanka'.format(neighborhood))
        coords = geo.latlng
    return coords

coords = [ get_latlng(neighborhood) for neighborhood in col_df["Neighborhood"].tolist() ]

coords

[[6.871710000000064, 79.99736000000007],
 [6.904660000000035, 79.85480000000007],
 [6.905200000000036, 79.91554000000008],
 [6.856881411183795, 79.89399987308153],
 [6.954920000000072, 79.86670000000004],
 [6.840800000000058, 79.90441000000004],
 [6.915870000000041, 79.87760000000003],
 [6.9094800000000305, 79.86924000000005],
 [6.931940000000054, 79.84555000000006],
 [6.9705600000000345, 79.91224000000005],
 [6.851320000000044, 79.86590000000007],
 [6.840470000000039, 79.87824000000006],
 [6.937220000000025, 79.88221000000004],
 [6.94148000000007, 79.84664000000004],
 [6.946550000000059, 79.87034000000006],
 [6.887580000000071, 79.86255000000006],
 [6.876463499999996, 79.935723],
 [6.936210000000074, 79.85844000000003],
 [6.931940000000054, 79.84555000000006],
 [6.856959882086081, 79.87853264931098],
 [6.935580000000073, 79.98416000000003],
 [6.7816400000000385, 79.98748000000006],
 [6.866550000000075, 79.87646000000007],
 [6.842750000000024, 79.87202000000008],
 [6.98237000000006, 79

In [41]:
# dataframe to load the coordinates into Latitude and Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

col_df['Latitude'] = df_coords['Latitude']
col_df['Longitude'] = df_coords['Longitude']
print(col_df.shape)
col_df

(65, 3)


Neighborhood  Latitude  Longitude
0                  Athurugiriya  6.871710  79.997360
1                 Bambalapitiya  6.904660  79.854800
2                  Battaramulla  6.905200  79.915540
3                     Batuwatta  6.856881  79.894000
4                   Bloemendhal  6.954920  79.866700
5                 Boralesgamuwa  6.840800  79.904410
6                       Borella  6.915870  79.877600
7              Cinnamon Gardens  6.909480  79.869240
8                       Colombo  6.931940  79.845550
9                      Dalugama  6.970560  79.912240
10                     Dehiwala  6.851320  79.865900
11       Dehiwala-Mount Lavinia  6.840470  79.878240
12                   Dematagoda  6.937220  79.882210
13               Fort (Colombo)  6.941480  79.846640
14                    Grandpass  6.946550  79.870340
15                Havelock Town  6.887580  79.862550
16                    Hokandara  6.876463  79.935723
17                   Hulftsdorp  6.936210  79.858440
18                       Ja-Ela  6.931940  79.845550
19                    Kadawatha  6.856960  79.878533
20   Kaduwela, Western Province  6.935580  79.984160
21                  Kahathuduwa  6.781640  79.987480
22                   Kalubowila  6.866550  79.876460
23                      Kandana  6.842750  79.872020
24                 Kiribathgoda  6.982370  79.927540
25                   Kirulapana  6.878622  79.873219
26                     Kohuwala  6.869590  79.888900
27                  Kollupitiya  6.912690  79.851330
28                    Kolonnawa  6.932630  79.888860
29                     Koswatte  6.901920  79.894080
30                     Kotahena  6.959150  79.868070
31                  Kotikawatta  6.931380  79.916140
32                      Kottawa  6.841360  79.965200
33                  Madampitiya  6.963820  79.874970
34            Maha Nuge Gardens  6.916680  79.849062
35                   Maharagama  6.846740  79.927110
36                       Malabe  6.903980  79.955180
37                  Maligawatta  6.933660  79.871850
38                     Maradana  6.924730  79.865700
39                 Mattakkuliya  6.972300  79.875320
40                       Modara  6.966730  79.871410
41                     Moratuwa  6.778910  79.883100
42                Mount-Lavinia  6.839810  79.866830
43                  Narahenpita  6.898260  79.879810
44                       Nawala  6.893230  79.890030
45                     Nugegoda  6.877060  79.892920
46                    Pamankada  6.879540  79.869370
47                Panchikawatte  6.932410  79.863020
48                  Pannipitiya  6.845370  79.939010
49                    Pelawatte  6.893200  79.937190
50                   Peliyagoda  6.948670  79.896510
51            Pettah, Sri Lanka  6.938140  79.852110
52                  Piliyandala  6.801610  79.922710
53                       Ragama  6.931940  79.845550
54                   Rajagiriya  6.907290  79.895430
55                    Ratmalana  6.814470  79.877950
56                 Slave Island  6.926520  79.850200
57    Sri Jayawardenepura Kotte  6.883960  79.900800
58  Template:Suburbs of Colombo  6.931940  79.845550
59               Thalawathugoda  6.873810  79.943780
60                  Union Place  6.918760  79.861130
61                      Wattala  6.931940  79.845550
62                     Welikada  6.904536  79.895614
63                   Wellawatte  6.879240  79.859890
64            Wickramasinhapura  6.881652  79.937361

In [42]:
col_df.to_csv("col_df.csv", index=False)

Map of Colombo with neighbourhoods

In [45]:
address = 'Colombo, SriLanka'

geolocator = Nominatim(user_agent="my-application")
loc = geolocator.geocode(address)
lat = loc.latitude
long = loc.longitude
print('The geograpical coordinate of Colombo, SriLanka {}, {}.'.format(lat, long))

The geograpical coordinate of Colombo, SriLanka 6.9218124, 79.8655608840961.


In [48]:
map_col = folium.Map(location=[lat, long], zoom_start=11)

for lat, long, neighborhood in zip(col_df['Latitude'], col_df['Longitude'], col_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker([lat, long],radius=3,popup=label,color='blue',fill=True,fill_color='#3170cc',fill_opacity=0.6).add_to(map_col)  
    
map_col

In [49]:
map_col.save('map_col.html')

Explore Colombo, SriLanka neighbourhoods thourgh Foursquare API

In [78]:
CLIENT_ID = '3SQEYD1KWPOCNP0QL0DOZCEYGJYMXR0CLUFFATVYG5UUJESQ'
CLIENT_SECRET = 'IKKAY0J4Q4OLJGARA05CGN4UZMAR0IZSRJMFSV1GRSRZQ40Z'
VERSION = '20180605'

In [79]:
radius = 500
LIMIT = 100
venues = []
for lat,long,neighborhood in zip(col_df['Latitude'],col_df['Longitude'],col_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(CLIENT_ID,CLIENT_SECRET,
        VERSION,lat,long,radius, LIMIT)
    results = requests.get(url).json()["response"]['groups'][0]['items']
    for venue in results:
        venues.append((neighborhood,lat,long,venue['venue']['name'],venue['venue']['location']['lat'], venue['venue']['location']['lng'],
        venue['venue']['categories'][0]['name']))

In [80]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(997, 7)


Neighborhood  Latitude  Longitude  \
0   Athurugiriya   6.87171   79.99736   
1   Athurugiriya   6.87171   79.99736   
2   Athurugiriya   6.87171   79.99736   
3  Bambalapitiya   6.90466   79.85480   
4  Bambalapitiya   6.90466   79.85480   

                                       VenueName  VenueLatitude  \
0                               Vithanage Stores       6.870551   
1  Pasindu Auto Parts & P.A. Products (Pvt) Ltd.       6.873748   
2                               Walgama Junction       6.869385   
3                                 99X Technology       6.905620   
4                               The Cake Factory       6.905188   

   VenueLongitude      VenueCategory  
0       79.998906  Convenience Store  
1       79.994840      Auto Workshop  
2       79.994476       Intersection  
3       79.854946        IT Services  
4       79.856409       Dessert Shop

In [54]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Neighborhood                                                                 
Athurugiriya                        3          3          3              3   
Bambalapitiya                      40         40         40             40   
Battaramulla                       11         11         11             11   
Batuwatta                           7          7          7              7   
Boralesgamuwa                       5          5          5              5   
Borella                            16         16         16             16   
Cinnamon Gardens                   41         41         41             41   
Colombo                            51         51         51             51   
Dalugama                            2          2          2              2   
Dehiwala                           22         22         22             22   
Dehiwala-Mount Lavinia              3          3          3              3   
Dematagoda                          2          2          2              2   
Fort (Colombo)                      4          4          4              4   
Grandpass                           5          5          5              5   
Havelock Town                      39         39         39             39   
Hokandara                          12         12         12             12   
Hulftsdorp                          4          4          4              4   
Ja-Ela                             51         51         51             51   
Kadawatha                           6          6          6              6   
Kaduwela, Western Province         10         10         10             10   
Kalubowila                         10         10         10             10   
Kandana                             3          3          3              3   
Kiribathgoda                        7          7          7              7   
Kirulapana                         15         15         15             15   
Kohuwala                           28         28         28             28   
Kollupitiya                        59         59         59             59   
Kolonnawa                           4          4          4              4   
Koswatte                            7          7          7              7   
Kotahena                            2          2          2              2   
Kotikawatta                         3          3          3              3   
Kottawa                            16         16         16             16   
Madampitiya                         1          1          1              1   
Maha Nuge Gardens                  43         43         43             43   
Maharagama                         24         24         24             24   
Malabe                             11         11         11             11   
Maligawatta                         7          7          7              7   
Maradana                           15         15         15             15   
Mattakkuliya                        6          6          6              6   
Modara                              1          1          1              1   
Moratuwa                            4          4          4              4   
Mount-Lavinia                      35         35         35             35   
Narahenpita                         6          6          6              6   
Nawala                              8          8          8              8   
Nugegoda                           15         15         15             15   
Pamankada                          13         13         13             13   
Panchikawatte                      11         11         11             11   
Pannipitiya                         4          4          4              4   
Pelawatte                           2          2          2              2   
Pettah, Sri Lanka                  16         16         16             16   
Piliyandala                         4          4   

In [81]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 154 uniques categories.


In [82]:
#  All the categories
venues_df['VenueCategory'].unique()[:50]

array(['Convenience Store', 'Auto Workshop', 'Intersection',
       'IT Services', 'Dessert Shop', 'Restaurant', 'Seafood Restaurant',
       'Chinese Restaurant', 'Coffee Shop', 'Bakery', 'Boutique', 'Café',
       'Middle Eastern Restaurant', "Men's Store", 'Casino',
       'Fast Food Restaurant', 'Sandwich Place', 'Optical Shop', 'Office',
       'Jewelry Store', 'Bookstore', 'Grocery Store', 'Lingerie Store',
       'Sri Lankan Restaurant', 'Sporting Goods Shop', 'Nightclub',
       'Clothing Store', 'Japanese Restaurant', 'Bar', 'Food Court',
       'Electronics Store', 'Lake', 'Cocktail Bar', 'Department Store',
       'Italian Restaurant', 'Asian Restaurant', 'Mobile Phone Shop',
       'Supermarket', 'Playground', 'Print Shop', 'Shopping Mall',
       "Women's Store", 'Gym', 'Pizza Place', 'Hotel', 'Cosmetics Shop',
       'Movie Theater', 'Park', 'Vegetarian / Vegan Restaurant',
       'Theater'], dtype=object)

In [83]:
# check if restaurant category is there"
"Restaurant" in venues_df['VenueCategory'].unique()

True

In [84]:
colombo_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

colombo_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

fixed_columns = list(colombo_onehot.columns[-1:]) + list(colombo_onehot.columns[:-1])
colombo_onehot = colombo_onehot[fixed_columns]

print(colombo_onehot.shape)
colombo_onehot.head()

(997, 155)


Neighborhoods  American Restaurant  Arcade  Art Gallery  \
0   Athurugiriya                    0       0            0   
1   Athurugiriya                    0       0            0   
2   Athurugiriya                    0       0            0   
3  Bambalapitiya                    0       0            0   
4  Bambalapitiya                    0       0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              1   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bakery  Bank  Bar  Beach  Beach Bar  \
0          0           0       0     0    0      0          0   
1          0           0       0     0    0      0          0   
2          0           0       0     0    0      0          0   
3          0           0       0     0    0      0          0   
4          0           0       0     0    0      0          0   

   Bed & Breakfast  Beer Garden  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                0            0       0              0          0         0   
1                0            0       0              0          0         0   
2                0            0       0              0          0         0   
3                0            0       0              0          0         0   
4                0            0       0              0          0         0   

   Bowling Alley  Breakfast Spot  Bubble Tea Shop  Buffet  Burger Joint  \
0              0               0                0       0             0   
1              0               0                0       0             0   
2              0               0                0       0             0   
3              0               0                0       0             0   
4              0               0                0       0             0   

   Bus Line  Bus Station  Bus Stop  Cafeteria  Café  Camera Store  \
0         0            0         0          0     0             0   
1         0            0         0          0     0             0   
2         0            0         0          0     0             0   
3         0            0         0          0     0             0   
4         0            0         0          0     0             0   

   Candy Store  Casino  Chinese Restaurant  Clothing Store  Cocktail Bar  \
0            0       0                   0               0             0   
1            0       0                   0               0             0   
2            0       0                   0               0             0   
3            0       0                   0               0             0   
4            0       0                   0               0             0   

   Coffee Shop  Concert Hall  Convenience Store  Cosmetics Shop  \
0            0             0                  1               0   
1            0             0                  0               0   
2            0             0                  0               0   
3            0             0                  0               0   
4            0             0                  0               0   

   Cricket Ground  Dance Studio  Department Store  Design Studio  \
0               0             0                 0              0   
1               0             0                 0              0   
2               0             0                 0              0   
3               0             0                 0              0   
4               0             0                 0              0   

   Dessert Shop  Diner  Donut Shop  Eastern European Restaurant  \
0             0      0           0                            0   
1             0     

In [85]:
col_com = colombo_onehot.groupby(["Neighborhoods"]).mean().reset_index()
print(col_com.shape)
col_com

(62, 155)


Neighborhoods  American Restaurant    Arcade  Art Gallery  \
0                  Athurugiriya             0.000000  0.000000     0.000000   
1                 Bambalapitiya             0.000000  0.000000     0.000000   
2                  Battaramulla             0.000000  0.000000     0.000000   
3                     Batuwatta             0.000000  0.000000     0.000000   
4                 Boralesgamuwa             0.000000  0.000000     0.000000   
5                       Borella             0.000000  0.000000     0.000000   
6              Cinnamon Gardens             0.000000  0.000000     0.048780   
7                       Colombo             0.000000  0.000000     0.000000   
8                      Dalugama             0.000000  0.000000     0.000000   
9                      Dehiwala             0.000000  0.000000     0.000000   
10       Dehiwala-Mount Lavinia             0.000000  0.000000     0.000000   
11                   Dematagoda             0.000000  0.000000     0.000000   
12               Fort (Colombo)             0.000000  0.000000     0.000000   
13                    Grandpass             0.000000  0.000000     0.000000   
14                Havelock Town             0.000000  0.000000     0.000000   
15                    Hokandara             0.000000  0.000000     0.000000   
16                   Hulftsdorp             0.000000  0.000000     0.000000   
17                       Ja-Ela             0.000000  0.000000     0.000000   
18                    Kadawatha             0.000000  0.000000     0.000000   
19   Kaduwela, Western Province             0.000000  0.000000     0.000000   
20                   Kalubowila             0.000000  0.000000     0.000000   
21                      Kandana             0.000000  0.000000     0.000000   
22                 Kiribathgoda             0.000000  0.000000     0.000000   
23                   Kirulapana             0.000000  0.000000     0.066667   
24                     Kohuwala             0.000000  0.000000     0.000000   
25                  Kollupitiya             0.000000  0.000000     0.000000   
26                    Kolonnawa             0.000000  0.000000     0.000000   
27                     Koswatte             0.000000  0.000000     0.000000   
28                     Kotahena             0.000000  0.000000     0.000000   
29                  Kotikawatta             0.000000  0.000000     0.000000   
30                      Kottawa             0.000000  0.000000     0.000000   
31                  Madampitiya             0.000000  0.000000     0.000000   
32            Maha Nuge Gardens             0.000000  0.000000     0.000000   
33                   Maharagama             0.000000  0.000000     0.000000   
34                       Malabe             0.000000  0.000000     0.000000   
35                  Maligawatta             0.000000  0.000000     0.000000   
36                     Maradana             0.000000  0.000000     0.000000   
37                 Mattakkuliya             0.000000  0.000000     0.000000   
38                       Modara             0.000000  0.000000     0.000000   
39                     Moratuwa             0.000000  0.000000     0.000000   
40                Mount-Lavinia             0.000000  0.028571     0.000000   
41                  Narahenpita             0.000000  0.000000     0.000000   
42                       Nawala             0.000000  0.000000     0.000000   
43                     Nugegoda             0.000000  0.000000     0.000000   
44                    Pamankada             0.000000  0.000000     0.076923   
45                Panchikawatte             0.000000  0.000000     0.000000   
46                  Pannipitiya             0.000000  0.000000     0.000000   
47                    Pelawatte             0.000000  0.000000     0.500000   
48            Pettah, Sri Lanka             0.000000  0.000000     0.000000   
49                  Piliyandala             0.000000  0.000000     0

In [86]:
len(col_com[col_com["Restaurant"] > 0])

25

In [87]:
col_rest = col_com[["Neighborhoods","Restaurant"]]
col_rest.head()

Neighborhoods  Restaurant
0   Athurugiriya        0.00
1  Bambalapitiya        0.05
2   Battaramulla        0.00
3      Batuwatta        0.00
4  Boralesgamuwa        0.20

k-means clustering to cluster the neighborhoods nto 3 clusters

In [88]:
kclusters = 3

col_clustering = col_com.drop(["Neighborhoods"], 1)
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(col_clustering)
kmeans.labels_[0:10]

array([1, 1, 1, 0, 0, 1, 1, 1, 1, 1], dtype=int32)

In [89]:
col_group = col_rest.copy()
col_group["Cluster Labels"] = kmeans.labels_
col_group.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
col_group = col_group.join(col_df.set_index("Neighborhood"), on="Neighborhood")
print(col_group.shape)
col_group.head()

(62, 5)


Neighborhood  Restaurant  Cluster Labels  Latitude  Longitude
0   Athurugiriya        0.00               1  6.871710   79.99736
1  Bambalapitiya        0.05               1  6.904660   79.85480
2   Battaramulla        0.00               1  6.905200   79.91554
3      Batuwatta        0.00               0  6.856881   79.89400
4  Boralesgamuwa        0.20               0  6.840800   79.90441

In [90]:
col_group.sort_values(["Cluster Labels"], inplace=True)
col_group

Neighborhood  Restaurant  Cluster Labels  Latitude  \
54    Sri Jayawardenepura Kotte    0.000000               0  6.883960   
28                     Kotahena    0.000000               0  6.959150   
3                     Batuwatta    0.000000               0  6.856881   
4                 Boralesgamuwa    0.200000               0  6.840800   
27                     Koswatte    0.142857               0  6.901920   
26                    Kolonnawa    0.000000               0  6.932630   
41                  Narahenpita    0.166667               0  6.898260   
52                    Ratmalana    0.000000               0  6.814470   
42                       Nawala    0.125000               1  6.893230   
40                Mount-Lavinia    0.171429               1  6.839810   
39                     Moratuwa    0.000000               1  6.778910   
0                  Athurugiriya    0.000000               1  6.871710   
43                     Nugegoda    0.066667               1  6.877060   
36                     Maradana    0.000000               1  6.924730   
35                  Maligawatta    0.142857               1  6.933660   
37                 Mattakkuliya    0.000000               1  6.972300   
44                    Pamankada    0.000000               1  6.879540   
47                    Pelawatte    0.000000               1  6.893200   
46                  Pannipitiya    0.000000               1  6.845370   
34                       Malabe    0.000000               1  6.903980   
48            Pettah, Sri Lanka    0.000000               1  6.938140   
49                  Piliyandala    0.000000               1  6.801610   
50                       Ragama    0.039216               1  6.931940   
51                   Rajagiriya    0.062500               1  6.907290   
53                 Slave Island    0.000000               1  6.926520   
55  Template:Suburbs of Colombo    0.039216               1  6.931940   
56               Thalawathugoda    0.000000               1  6.873810   
57                  Union Place    0.111111               1  6.918760   
58                      Wattala    0.039216               1  6.931940   
59                     Welikada    0.000000               1  6.904536   
45                Panchikawatte    0.000000               1  6.932410   
33                   Maharagama    0.000000               1  6.846740   
30                      Kottawa    0.125000               1  6.841360   
31                  Madampitiya    0.000000               1  6.963820   
1                 Bambalapitiya    0.050000               1  6.904660   
2                  Battaramulla    0.000000               1  6.905200   
5                       Borella    0.062500               1  6.915870   
6              Cinnamon Gardens    0.048780               1  6.909480   
7                       Colombo    0.039216               1  6.931940   
8                      Dalugama    0.000000               1  6.970560   
9                      Dehiwala    0.000000               1  6.851320   
10       Dehiwala-Mount Lavinia    0.000000               1  6.840470   
11                   Dematagoda    0.000000               1  6.937220   
12               Fort (Colombo)    0.000000               1  6.941480   
13                    Grandpass    0.000000               1  6.946550   
14                Havelock Town    0.025641               1  6.887580   
32            Maha Nuge Gardens    0.046512               1  6.916680   
15                    Hokandara    0.083333               1  6.876463   
17                       Ja-Ela    0.039216               1  6.931940   
18                    Kadawatha    0.000000               1  6.856960   
19   Kaduwela, Western Province    0.000000               1  6.935580   
20                   Kalubowila    0.000000               1  6.866550   
21                      Kandana    0.333333               1  6.842750   
22                 Kiribathgoda    0.000000               1  6.982370   
23  

In [91]:
map_clusters = folium.Map(location=[lat, long], zoom_start=11)

x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(col_group['Latitude'], col_group['Longitude'], col_group['Neighborhood'], col_group['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker([lat, lon],radius=3,popup=label,color=rainbow[cluster-1],fill=True,fill_color=rainbow[cluster-1],
                        fill_opacity=0.6).add_to(map_clusters)
       
map_clusters

In [92]:
map_clusters.save('map_clusters.html')

Verify clusters

In [93]:
col_group.loc[col_group['Cluster Labels'] == 0]

Neighborhood  Restaurant  Cluster Labels  Latitude  Longitude
54  Sri Jayawardenepura Kotte    0.000000               0  6.883960   79.90080
28                   Kotahena    0.000000               0  6.959150   79.86807
3                   Batuwatta    0.000000               0  6.856881   79.89400
4               Boralesgamuwa    0.200000               0  6.840800   79.90441
27                   Koswatte    0.142857               0  6.901920   79.89408
26                  Kolonnawa    0.000000               0  6.932630   79.88886
41                Narahenpita    0.166667               0  6.898260   79.87981
52                  Ratmalana    0.000000               0  6.814470   79.87795

In [94]:
col_group.loc[col_group['Cluster Labels'] == 1]

Neighborhood  Restaurant  Cluster Labels  Latitude  \
42                       Nawala    0.125000               1  6.893230   
40                Mount-Lavinia    0.171429               1  6.839810   
39                     Moratuwa    0.000000               1  6.778910   
0                  Athurugiriya    0.000000               1  6.871710   
43                     Nugegoda    0.066667               1  6.877060   
36                     Maradana    0.000000               1  6.924730   
35                  Maligawatta    0.142857               1  6.933660   
37                 Mattakkuliya    0.000000               1  6.972300   
44                    Pamankada    0.000000               1  6.879540   
47                    Pelawatte    0.000000               1  6.893200   
46                  Pannipitiya    0.000000               1  6.845370   
34                       Malabe    0.000000               1  6.903980   
48            Pettah, Sri Lanka    0.000000               1  6.938140   
49                  Piliyandala    0.000000               1  6.801610   
50                       Ragama    0.039216               1  6.931940   
51                   Rajagiriya    0.062500               1  6.907290   
53                 Slave Island    0.000000               1  6.926520   
55  Template:Suburbs of Colombo    0.039216               1  6.931940   
56               Thalawathugoda    0.000000               1  6.873810   
57                  Union Place    0.111111               1  6.918760   
58                      Wattala    0.039216               1  6.931940   
59                     Welikada    0.000000               1  6.904536   
45                Panchikawatte    0.000000               1  6.932410   
33                   Maharagama    0.000000               1  6.846740   
30                      Kottawa    0.125000               1  6.841360   
31                  Madampitiya    0.000000               1  6.963820   
1                 Bambalapitiya    0.050000               1  6.904660   
2                  Battaramulla    0.000000               1  6.905200   
5                       Borella    0.062500               1  6.915870   
6              Cinnamon Gardens    0.048780               1  6.909480   
7                       Colombo    0.039216               1  6.931940   
8                      Dalugama    0.000000               1  6.970560   
9                      Dehiwala    0.000000               1  6.851320   
10       Dehiwala-Mount Lavinia    0.000000               1  6.840470   
11                   Dematagoda    0.000000               1  6.937220   
12               Fort (Colombo)    0.000000               1  6.941480   
13                    Grandpass    0.000000               1  6.946550   
14                Havelock Town    0.025641               1  6.887580   
32            Maha Nuge Gardens    0.046512               1  6.916680   
15                    Hokandara    0.083333               1  6.876463   
17                       Ja-Ela    0.039216               1  6.931940   
18                    Kadawatha    0.000000               1  6.856960   
19   Kaduwela, Western Province    0.000000               1  6.935580   
20                   Kalubowila    0.000000               1  6.866550   
21                      Kandana    0.333333               1  6.842750   
22                 Kiribathgoda    0.000000               1  6.982370   
23                   Kirulapana    0.000000               1  6.878622   
24                     Kohuwala    0.107143               1  6.869590   
25                  Kollupitiya    0.050847               1  6.912690   
29                  Kotikawatta    0.000000               1  6.931380   
60                   Wellawatte    0.000000               1  6.879240   
16                   Hulftsdorp    0.000000               1  6.936210   
61            Wickramasinhapura    0.250000               1  6.881652   

    Longitude  
42  79.890030  
40  79.866830  
39  79.883100  
0   79.99736

In [95]:
col_group.loc[col_group['Cluster Labels'] == 2]

Neighborhood  Restaurant  Cluster Labels  Latitude  Longitude
38       Modara         0.0               2   6.96673   79.87141

## Conclusion

As per the observation, almost 85% of restaurants are in second Cluster area of Colombo and approx 13% in frst cluster. Instead third cluster has only one restaurant in the neighborhood. This presents a great business potential to open new restaurants in third cluster as there is negligible competition from other restaurants and also people need not go to far area restaurants. Restaurants in second cluster will be going through intense competition due to high concentration of restaurants. It shows that there are surplus of restaurants in a single neighbourhood of the city, with other neighbourhoods still having very few restaurants. Hence, this analysis recommends restaurant owners to take advantage of these findings to open new restaurants in neighborhoods in third cluster to expand their business. Restaurants with exclusive cuisines who stand out from the competition can also open new restaurants in neighborhoods in first cluster with moderate competition. Lastly, restaurant owners are advised to shun neighborhoods in second cluster that already has high number of restaurats and going through intense competition.